### DISCUSSION TITLE
How can i generate a static token for a user?

### DISCUSSION THREAD

**Discussion by niklasgrewe** [directus/directus#2822] [Archive]
Hi, i found this from the docs:

> Since static tokens don't change/rotate, they are inherently less secure than authenticating with user credentials and JWT. Therefore we do not expose them directly through the Directus App. For now, to set/change a user's static token, you must do so directly in the database at directus_users.token (plaintext).

https://docs.directus.io/guides/users.html#creating-users

But how can i create a static token, so i can use it in the Javascript SDK? I honestly have no idea what exactly I need to do here... Which steps must be taken?

And after that, which `mode` do I have to specify in the JS SDK? `jwt` or `cookie`? When i try this:

```js
const client = new DirectusSDK({
  url: "https://api.directus.cloud/",
  project: "dcABCdefHIJklm",
  token: "demo"
});
```
i get the typescript error:
```
Argument of type '{ url: string; project: string; token: string; }' is not assignable to parameter of type 'IConfigurationOptions'.
```
as it looks like i always have to specify the `mode` when i work in typescript, right or how can i solve this typescript error?

**Comment by rijkvanzanten** [comment]
The static token is saved under `directus_users.token`. You can add that in the DB or through the API.

> And after that, which mode do I have to specify in the JS SDK? jwt or cookie?

It depends on what your use case is. If this is a front-end (browser) website/app you're building, I'd recommend using the cookie mode as that's the only safe way to store tokens on a browser environment.

> as it looks like i always have to specify the mode when i work in typescript, right 

That's correct

---

The mode flag tells the API whether to return the access token as a JWT token (string), or set it in a cookie on login. The cookie mode is the safest for browser use, but might not always be the correct option (for example when using server to server communication).

**Reply by rijkvanzanten** [reply]
Using static tokens on a front-end project is not recommended by the way. By doing that, you're effectively publicizing your static token, and thus giving the world access to that user. If you're looking for that sort of public data, I'd recommend opening up the permissions for the public role instead.

**Reply by niklasgrewe** [reply]
@rijkvanzanten 
> You can add that in the DB or through the API

how can i do that with the api? sorry i am newcomer of directus 😅

> The cookie mode is the safest for browser use

does the sdk use the http-only cookie, so xss attacks are prevented?

> Using static tokens on a front-end project is not recommended by the way.

i want to use the token on server side 😌

- - -

i want to use the sdk server and client-side in parallel... what is the best way to do this? I am using Sapper and Svelte in SSR (server-side-rendering) Mode with Express instead of polka. I need the sdk server-side to get some data and send emails with nodemailer. On the client-side i need the sdk to login user, store data etc. 

how can I use the sdk in this case? Does it make sense to use the middleware or should I import the sdk twice, once server and once client?
which auth models should I use in both cases? static token server side and what should I use on the client side? username and password would be the worst, i think 😅

can i hide the directus url on the client side? i do not necessarily want to "publish" this...

**Comment by alexvdvalk** [comment]
Updated link to documentation: https://docs.directus.io/app/user-directory.html#admin-options

Static tokens are managed in your user page in V9
